File: comparisons/contapalavras/ContaPalavras.c
/* gcc ContaPalavras.c -o ContaPalavras.exe */

#include <stdio.h>
#include <ctype.h>


int contarPalavras(const char * frase) {
	int quantidadePalavras = 0;
	int posicaoCaracter;
	char caracterAnterior = ' ';
	char caracterAtual;

	for(posicaoCaracter = 0; caracterAtual = frase[posicaoCaracter], caracterAtual != '\0'; posicaoCaracter++) {

		if(!isalpha(caracterAnterior) && isalpha(caracterAtual))
			quantidadePalavras++;

		caracterAnterior = caracterAtual;
	}

	return quantidadePalavras;
}

int main() {
	char frase[1024];
	int quantidadePalavras;

	printf("Digite uma frase\n");

	/* http://stackoverflow.com/questions/6282198/reading-string-from-input-with-space-character */
	scanf("%[^\n]s", frase); 
 
	quantidadePalavras = contarPalavras(frase);
 
	printf("A frase tem %d palavras.\n", quantidadePalavras);

	return 0;
}

File: comparisons/contapalavras/ContaPalavras.cs
using System;
using System.Linq;


class Program
{
    static void Main(string[] args)
    {
        Console.WriteLine("Digite uma frase.");
        var frase = Console.ReadLine();
        var quantidadePalavras = ContaPalavras(frase);
        Console.WriteLine("A frase tem {0} palavreas", quantidadePalavras);
    }

    private static int ContaPalavras(string frase)
    {
        return
            frase
                .Aggregate(
                    " "
                    , (acc, x) => char.IsLetter(acc.Last()) && char.IsLetter(x) ? acc : acc += x.ToString()
                    , acc => acc.Count(x => char.IsLetter(x))
                );
    }
}
